src/Features/CSharp/Portable/UseCompoundAssignment/CSharpUseCompoundCoalesceAssignmentDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.UseCompoundAssignment;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCompoundAssignment
+{
+    /// <summary>
+    /// Looks for expressions of the form <c>expr ?? (expr = value)</c> and converts it to
+    /// <c>expr ??= value</c>.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpUseCompoundCoalesceAssignmentDiagnosticAnalyzer
+        : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpUseCompoundCoalesceAssignmentDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UseCoalesceCompoundAssignmentDiagnosticId,
+                   CodeStyleOptions.PreferCompoundAssignment,
+                   new LocalizableResourceString(nameof(FeaturesResources.Use_compound_assignment), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeCoalesceExpression, SyntaxKind.CoalesceExpression);
+
+        private void AnalyzeCoalesceExpression(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = context.SemanticModel;
+            var coalesceExpression = (BinaryExpressionSyntax)context.Node;
+
+            var option = context.GetOption(CodeStyleOptions.PreferCompoundAssignment, coalesceExpression.Language);
+
+            // Bail immediately if the user has disabled this feature.
+            if (!option.Value)
+                return;
+
+            var coalesceLeft = coalesceExpression.Left;
+            var coalesceReft = coalesceExpression.Right;
+
+            if (!(coalesceReft is ParenthesizedExpressionSyntax parenthesizedExpr))
+                return;
+
+            if (!(parenthesizedExpr.Expression is AssignmentExpressionSyntax assignment))
+                return;
+
+            if (assignment.Kind() != SyntaxKind.SimpleAssignmentExpression)
+                return;
+
+            // have    x ?? (y = z)
+            // ensure that 'x' and 'y' are suitably equivalent.
+            var syntaxFacts = CSharpSyntaxFacts.Instance;
+            if (!syntaxFacts.AreEquivalent(coalesceLeft, assignment.Left))
+                return;
+
+            // Syntactically looks promising.  But we can only safely do this if 'expr'
+            // is side-effect-free since we will be changing the number of times it is
+            // executed from twice to once.
+            if (!UseCompoundAssignmentUtilities.IsSideEffectFree(
```

Comment:
Tagging @333fred in case he know of any special cases that would need to be considered when converting from `??` to `??=`